library/std/src/sys/sync/rwlock/queue.rs

```diff
@@ -456,25 +484,83 @@ impl RwLock {
         loop {
             // Atomically release the lock and try to acquire the queue lock.
-            let next = state.map_addr(|a| (a & !LOCKED) | QUEUE_LOCKED);
-            match self.state.compare_exchange_weak(state, next, AcqRel, Relaxed) {
-                // The queue lock was acquired. Release it, waking up the next
-                // waiter in the process.
-                Ok(_) if state.addr() & QUEUE_LOCKED == 0 => unsafe {
-                    return self.unlock_queue(next);
-                },
+            let next = state.wrapping_byte_sub(LOCKED);
+            if state.addr() & QUEUE_LOCKED != 0 {
                 // Another thread already holds the queue lock, leave waking up
                 // waiters to it.
-                Ok(_) => return,
-                Err(new) => state = new,
+                match self.state.compare_exchange_weak(state, next, Release, Relaxed) {
+                    Ok(_) => return,
+                    Err(new) => state = new,
+                }
+            } else {
+                // Acquire the queue lock and wake up the next waiter.
+                let next = next.wrapping_byte_add(QUEUE_LOCKED);
+                match self.state.compare_exchange_weak(state, next, AcqRel, Relaxed) {
+                    Ok(_) => unsafe {
+                        return self.unlock_queue(next);
+                    },
+                    Err(new) => state = new,
+                }
             }
         }
     }
 
-    /// Unlocks the queue. If the lock is unlocked, wakes up the next eligible
-    /// thread(s).
+    #[inline]
+    pub unsafe fn downgrade(&self) {
+        // Atomically set to read-locked with a single reader, without any
+        // waiting threads.
+        if let Err(state) = self.state.compare_exchange(
+            without_provenance_mut(LOCKED),
+            without_provenance_mut(SINGLE | LOCKED),
+            Release,
+            Relaxed,
+        ) {
+            // The only way the state can have changed is if there are threads
+            // queued. Wake all of them up.
+            unsafe { self.downgrade_slow(state) }
+        }
+    }
+
+    #[cold]
+    unsafe fn downgrade_slow(&self, mut state: State) {
+        debug_assert!(state.addr() & (DOWNGRADE | QUEUED | LOCKED) == (QUEUED | LOCKED));
+        // Attempt to grab the entire waiter queue.
+        loop {
+            if state.addr() & QUEUE_LOCKED != 0 {
+                // Another thread already holds the queue lock. Tell it to wake
+                // up all waiters. If it completes before we release our lock,
+                // the effect will be just the same as if we had changed the
+                // state below. Otherwise, the `DOWNGRADE` bit will still be
+                // set, meaning `read_unlock` will realize that the lock is still
+                // exclusively locked and act accordingly.
+                let next = state.wrapping_byte_add(DOWNGRADE);
```

Comment:
Any reason why for non-atomic pointers, `wrapping_byte_add` / `wrapping_byte_sub` is used rather than `state.map_addr(|a| a | DOWNGRADE)` that shows up a few times? (relevant a few times in this file). I think mapping with `|` makes it a bit more clear that the goal is to set the bit rather than perform some arithmetic.

Reply:
I've changed them in 469dc95213db753fb6e810e6479bc03179a9a3e0